### PR TITLE
Add a utility to lay out actors in a grid

### PR DESCRIPTION
**Pull request by stevenvergenz**
I must've implemented this half a dozen times by now. Finally I have a grid layout that's reusable!

### CODE REVIEW

**Comment by stevenvergenz** [comment]
I'm surprised to hear this, I thought tables were pretty self-explanatory, with pre-defined expectations. Is this PR missing features that are so badly needed that shipping without the feature is worse than not having the layout at all?

**Comment by eanders-MS** [comment]
A grid layout utility is worthwhile addition. If it is going to be an SDK component, it just needs a design review. Some ideas I'd like to cover:
* Ability to arrange the grid on a surface of a sphere, cylinder, or plane (or on the surface of any primitive?)
* Ability to layout the grid tiled or radial.
* How the objects should orient in relation to their surface position (none, face-center, face-center-reversed, face-parent, etc.).

MRTK has something like this. See: https://microsoft.github.io/MixedRealityToolkit-Unity/Documentation/README_ObjectCollection.html

**Comment by stevenvergenz** [comment]
@eanders-MS These other modes do seem useful, but their lack shouldn't disqualify this PR I think. Perhaps I could simply rename the utility `PlanarGridLayout`?

**Comment by eanders-MS** [comment]
You're suggesting separate classes for different surface types? `SphericalGridLayout`, etc? I'm fine with that design. As we add more functionality, a singular layout class trying to be all things would eventually grow cumbersome.

File: packages/sdk/src/util/gridLayout.ts [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+			const cellPosition = new Vector3(
+				colWidths.slice(0, cell.column).reduce(sum, 0),
+				-rowHeights.slice(0, cell.row).reduce(sum, 0),
+				cell.contents.transform.local.position.z);
+			const cellAlign = GridLayout.getOffsetFromAlignment(
+				cell.alignment ?? this.defaultCellAlignment,
+				colWidths[cell.column], rowHeights[cell.row]
+			);
+
+			cell.contents.transform.local.position = gridAlign.add(cellPosition).add(cellAlign);
```
Comment (eanders-MS) [inline packages/sdk/src/util/gridLayout.ts]: >cell.contents.transform.local.position [](start = 3, length = 38)

Would be nice: Option to interpolate this change.

Comment (stevenvergenz) [reply packages/sdk/src/util/gridLayout.ts]: Easy to add, done!

**Comment by stevenvergenz** [comment]
I've added a design document, and made some naming tweaks. Review?

File: packages/functional-tests/src/tests/actor-spam-test.ts [hunk 2]
```diff
@@ -128,24 +132,22 @@ export default class ActorSpamTest extends Test {
...
-						local: {
-							position: {
-								x: -1 + spacing * (i % 25),
-								y: 0 - spacing * Math.floor(i / 25)
-							}
-						}
-					}
-				}
+					appearance: { meshId: ball.id }
+				}})
```
Comment (eanders-MS) [inline packages/functional-tests/src/tests/actor-spam-test.ts]: semicolon :)

File: packages/sdk/src/util/planarGridLayout.ts [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+	 * @param defaultCellAlignment How cells should be aligned by default.
+	 */
+	public constructor(
+		private anchor: Actor,
+		public gridAlignment = BoxAlignment.MiddleCenter,
+		public defaultCellAlignment = BoxAlignment.MiddleCenter
+	) { }
+
+	/** The number of columns in this grid. */
+	public columnCount() {
```
Comment (eanders-MS) [inline packages/sdk/src/util/planarGridLayout.ts]: >columnCount [](start = 8, length = 11)

I suggest making this and the rest getters. Or, prefix the method name with `get` --> `getColumnCount`

File: packages/sdk/src/util/planarGridLayout.ts [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+			height += this.rowHeight(i);
+		}
+		return height;
+	}
+
+	/**
+	 * The width of a particular column.
+	 * @param i The column index.
+	 */
+	public columnWidth(i: number) {
```
Comment (eanders-MS) [inline packages/sdk/src/util/planarGridLayout.ts]: >columnWidth [](start = 8, length = 11)

Suggest naming this `getColumnWidth`

File: packages/functional-tests/src/tests/grid-test.ts [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+			buttonGrid.addCell({
+				row: Math.floor(i / 3),
+				column: i % 3,
+				width: 0.3,
+				height: 0.3,
+				alignment,
+				contents: label
+			});
+		}
+		buttonGrid.applyLayout();
```
Comment (eanders-MS) [inline packages/functional-tests/src/tests/grid-test.ts]: >applyLayout [](start = 13, length = 11)

How about making this animate?

File: packages/functional-tests/src/tests/actor-spam-test.ts (hunk 2 above)
Comment (stevenvergenz) [reply packages/functional-tests/src/tests/actor-spam-test.ts]: This is still mid-statement, can't put a semi-colon here.

---
In reply to: [385870701](https://github.com/microsoft/mixed-reality-extension-sdk/pull/492#discussion_r385870701) [](ancestors = 385870701)

Comment (eanders-MS) [reply packages/functional-tests/src/tests/actor-spam-test.ts]: Oops I misread it.

File: packages/functional-tests/src/tests/grid-test.ts (hunk 5 above)
Comment (stevenvergenz) [reply packages/functional-tests/src/tests/grid-test.ts]: Why not?

---
In reply to: [385873835](https://github.com/microsoft/mixed-reality-extension-sdk/pull/492#discussion_r385873835) [](ancestors = 385873835)

**Pull request merged**